### ISSUE TITLE
Debugging Toga app on Windows

### ISSUE THREAD

**Issue by MicBoucinha** [beeware/toga#1048]
Hi everyone,

I heard of briefcase (0.3.3) and Toga (0.3.0.dev23) recently and decided to try it out. As such, I followed the Tutorial 0 on https://docs.beeware.org/en/latest/tutorial/tutorial-0.html

Everything seems to be working quite nicely from the command line and using the `briefcase dev` command the app runs perfectly fine.

I wanted to debug the app (vscode with the official Python extension) but so far I was unable to do so because I get a RuntimeError exception with the message `Toga application must have a formal name` when trying to launch the debugger over the module created. 

Do I need to try to attach the debugger to a running `briefcase dev` process?

Has anyone tried to debug the app on Windows, in vscode with the official Python extension that could point me in the right direction?

Thanks!

**Comment by freakboy3742** [comment]
I've done extensive debugging with pdb; so debugging is *possible*; however, I haven't tried setting up a VSCode debugging configuration (or any other IDE for that matter).

If it helps, the `Toga application must have a formal name` means that the app can't find either the `.dist-info` file for the module, or identify the name of the module that was started. If you've tried to start the app *before* running `briefcase dev`, the `.dist-info` file may not have been created, which could be the source of the problem.

If you *do* manage to sort out how to configure a debugging session, we'd definitely be interested in having that as an addition to the docs.

**Comment by MicBoucinha** [comment]
Hi,

Thanks for the tip, I will look into it. I ran `briefcase dev` before (and the `.dist-info`) but somehow the debugger gives the error nonetheless. It might be an issue with the working folder I am using.

In the meantime, I tried launching `briefcase dev` and afterwards, from vscode I used the debugger with the `Attach to process id` option and set a breakpoint in the callback of a Button and it worked nicely.

It would be easier to have the other way instead of attaching (less steps). I will continue to investigate how to make it work and post here my findings.

Thanks!

**Comment by KristofHorvath** [comment]
Hi,

I could launch the toga app from VSCode with the following lines in launch.json:

```
{
    "name": "Python: Module",
    "type": "python",
    "request": "launch",
    "module": "helloworld",
    "cwd": "${workspaceFolder}\\helloworld\\src"
}
```

However, breakpoints don't work correctly in this case. Only one in the main function is activated, but Step Into doesn't work on HelloWorld() call. All the other breakpoints are useless.

Have you got any idea?

**Comment by KristofHorvath** [comment]
Interesting update: built-in function breakpoint() works, and the program can be run step-by-step after it, although it's not a comfortable solution at all.

**Comment by MicBoucinha** [comment]
I was able to finally get back to this and @KristofHorvath's suggestion for the launch configuration works flawlessly for me with breakpoints either in the `startup` method and in the event handlers on a Mac and Python 3.8.1. 

However I keep getting the same `Toga application must have a formal name` exception on Windows.

@KristofHorvath from your cwd path I gather that you are on Windows. Which Python version are you running?

Thanks!

**Comment by KristofHorvath** [comment]
Yes, as you guessed, I am using Windows. I run the program on Python 3.8.0 in a virtual environment.

**Comment by MicBoucinha** [comment]
I must have some weird configuration on my Windows machine. I keep getting the exception.

I will keep trying to understand the issue and will post here if I manage to find a solution.

Thanks!

**Comment by meichthys** [comment]
@MicBoucinha I cannot seem to get this to work on MacOS - I also get the `Toga application must have a formal name` error when using @KristofHorvath 's launch.json

**Comment by MicBoucinha** [comment]
I wasn't able to make it work on Windows directly (only with `Attach to process`) and didn't try on Mac further than on my last comment, sorry.

**Comment by meichthys** [comment]
Ok, no problem, I'll work with pdb for now then.

**Comment by freakboy3742** [comment]
Closing as the question appears to have been resolved.

**Comment by peterpsc** [comment]
Not Resolved!  I have the same issue

**Comment by peterpsc** [comment]
Is this the way it should look???

import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW

class HelloWorld(toga.App):

    def __init__(self, formal_name, app_id, app_name):
        super().__init__(formal_name=formal_name, app_id=app_id, app_name=app_name)

    def startup(self):
        main_box = toga.Box()

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()
        return

def main():
    return HelloWorld("Hello World", "helloworld", "com.example.helloworld")

if __name__ == "__main__":
    main().main_loop()

**Comment by freakboy3742** [comment]
> Not Resolved! I have the same issue This issue will scare away anyone who wants to try it out for the first time, since the Tutorial is BROKEN If a user like me wants to edit source code in an IDE like PyCharm: 

No, the tutorial is *not* broken - or, at least, you haven't provided enough evidence to demonstrate that it is broken.

All I can tell you is that the code you've provided works fine in my testing, on every platform I can test it on (including Windows). I also have countless examples of people who have been able to successfully run the tutorial as written.

I strongly suspect the problem you're having is that you're not following the instructions in the tutorial as provided. If nothing else, the code you've provided in your second comment isn't the code that the tutorial prompts you write - so you're at least partially "off script".

Using an IDE like PyCharm to *edit* the code is irrelevant - it's an editor; a way to change the content of plain text files. I suspect the problem is occurring because you're using PyCharm to *run* the code.

The Briefcase tutorial provides explicit instructions on how to run the code. Those instructions are *independent* of your choice of IDE. Open a terminal session, and enter the commands *as described in the tutorial*, and all the evidence I have says the tutorial will run. 

How do you run the code in PyCharm? I don't know. I don't use PyCharm, so I can't speak to how you need to configure PyCharm. The tutorial doesn't recommend you use PyCharm. If you're taking the commands listed in the tutorial and putting them into some PyCharm "run" dialog, and expecting them to work... I have no idea what will happen. Evidently it doesn't work. But then, the tutorial doesn't prompt you do this, either.

I'll also note that you've added this comment to the end of a 3 year old issue that is, to my eye, unrelated to the problem you're having. The original reporter explicitly said "Everything seems to be working quite nicely from the command line and using the briefcase dev command the app runs perfectly fine." They were having a problem with VSCode, not PyCharm; and were asking how to configure VSCode to be used as a *debugger*. A solution was found (use pdb), there wasn't anything particularly actionable from the perspective of Toga (the project to which this issue is attached), and there hadn't been any updates in a year - so the issue was closed. 

If you're having difficulty with the tutorial, we'd be more than happy to help you through those difficulties. However, I'd suggest:
1) Starting a fresh discussion somewhere more appropriate - [the BeeWare Discussion forum](https://github.com/beeware/beeware/discussions) or [Discord server](https://beeware.org/bee/chat/) would be a good place to start
2) Provide more details about *exactly* what it is you've done, and the *specific* problem you're seeing.

**Comment by codewithmichael** [comment]
This is the top search result for **_"Toga application must have a formal name"_**

While I don't want to necro an old thread, this is where people are likely to land looking for how to debug a Toga app, and the solutions are fairly simple, and hopefully the following will provide what they're looking for

---

_**Assumption:** The developer has already run `briefcase dev` as described in the tutorial documentation, which creates the necessary `helloworld.dist-info` directory_

---

### Issue

The _"formal name"_ error arises because Toga is looking for the aforementioned `*.dist-info` directory based on the app module's package name. But if the application isn't run as a module (most debuggers won't by default), the `__package__` definition is empty

### Solution

If you're working purely from the command line, the `pdb` debugging module has an option flag, `-m`, to run as module as illustrated in the first example below

If, however, you're using a more standard IDE (like PyCharm), it's probably (1) using `pydev` to support remote debugging and (2) running the application straight from `__main__.py` (even if you request to run as a module)

To support either run case, jump to the second suggestion below

---

**If you're using _pdb_:**
Simply `cd` to the `src` directory and run the `helloworld` module via `pdb` in module mode:
```sh
$ python -m pdb -m helloworld
```

---

**If you're running _PyCharm_ (or another IDE)**:
You need to add a few lines to your Toga app's `__main__.py` file immediately before `main().main_loop()` to give it a package name. This is because many IDEs run modules by simply running thier `__main__.py` file, losing the package name in the process

Your `__main__.py` file should look like this:
```python
from helloworld.app import main

if __name__ == "__main__":

    # For Toga app debugging
    import sys
    if not sys.modules["__main__"].__package__:
        import os
        sys.modules["__main__"].__package__ = os.path.basename(os.path.dirname(__file__))

    main().main_loop()
```

The above uses the path to the current file (`__main__.py`) to determine the name of the parent directory (`helloworld`, which Python uses as the module name) and assigns it to the `__main__` module's `__package__` property, allowing Toga to find the parent's sibling directory, `helloworld.dist-info`

The above block should allow you to use any IDE to run/debug your Toga app

**Comment by freakboy3742** [comment]
That solution will definitely work - but there are two other options.

One option is to configure your PYTHONPATH. The invocation:

    $ python -m pdb -m helloworld

will only work if you're currently in the `src` folder, and won't detect installed metadata. This isn't how Briefcase launches an app. Instead, start from your project folder (the folder that contains your `pyproject.toml`), and run:

    $ PYTHONPATH=$(pwd)/src python -m pdb -m helloworld

This ensures that the src folder is on your path, so the app metadata will be discoverable.

The other option - provide an application name when you build your app. Briefcase, by default, generates an app that uses the `.dist-info` file for app metadata. However, this metadata can be provided explicitly. Briefcase will create a main method that looks something like:
```
def main():
    return HelloWorld()
```

If you modify this to read:
```
def main():
    return HelloWorld("Hello World", "com.example.helloworld")
```
explicitly providing a formal name and an app bundle ID), the app will not complain about missing metadata.

However, I'd suggest the first form (fixing the PYTHONPATH) is preferable.

As an aside, if you're ever in a situation where Briefcase at the command line works, but PyCharm (or any other IDE) doesn't, the issue is almost certainly that the IDE isn't using the same execution environment as Briefcase. If you want to know exactly what Briefcase is doing, you can run it in super verbose mode with the `-vv` command line switch. This will output the exact command line and execution environment that is used for every command Briefcase is executing. For example:

```
$ briefcase dev -vv

[helloworld] Starting in dev mode...

>>> Running Command:
>>>     /Users/rkm/beeware/briefcase/venv3.10/bin/python3.10 -c 'import runpy, sys;sys.path.pop(0);sys.argv.extend([]);runpy.run_module("helloworld", run_name="__main__", alter_sys=True)'
>>> Working Directory:
>>>     /Users/rkm
>>> Environment Overrides:
>>>     PYTHONPATH=/Users/rkm/beeware/toga/local/helloworld/src
>>>     PYTHONUNBUFFERED=1
>>>     PYTHONDEVMODE=1
>>>     PYTHONUTF8=1
===========================================================================
```

**Comment by NMertsch** [comment]
Another potential cause for this: If you run `pip install .` to install the project directly, this might generate a directory `src/<your-project>.egg-info/` containing basic project metadata. Don't do this! [Use `pip` only to install third-party packages](https://tutorial.beeware.org/en/latest/tutorial/tutorial-7/#adding-a-package), and use `briefcase` to run and package your application.

tl;dr: Delete `src/<your-project>.egg-info/`, run `briefcase dev` again.

Explanation:
* `pyproject.toml` is present, so `pip` installs the package.
* `pyproject.toml` does not contain PEP 621 project metadata (section `project`), so `pip` uses default values: `Version: 0.0.0`, `Name: <your-project-dirname>`.
    * `pip` does NOT pick up the briefcase project metadata from the section `tool.briefcase`!
* `pip` creates `src/<your-package>.egg-info/` containing this minimal set of project metadata.
* `briefcase dev` creates `src/<your-package>.dist-info/` with the correct metadata, but still reads the metadata from `src/<your-package>.egg-info/`, so toga doesn't find the `Formal-Name` and raises an exception.

See also https://github.com/beeware/briefcase/issues/2583
